cdap-hbase-compat-base/src/main/java/co/cask/cdap/replication/TableUpdater.java

```diff
@@ -0,0 +1,123 @@
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Map;
+import java.util.Timer;
+import java.util.TimerTask;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Utility class for Replication Status Tool
+ */
+public abstract class TableUpdater {
+
+  private static final Logger LOG = LoggerFactory.getLogger(TableUpdater.class);
+  private ConcurrentHashMap<String, Long> cachedUpdates;
+  private Timer updateTimer;
+
+  public final byte[] rsID;
+  public final byte[] columnFamily;
+  public final byte[] rowType;
+  public final Configuration conf;
+  public boolean tableExists = false;
+
+  public TableUpdater(String rowType, final Configuration conf) {
+    this.columnFamily = Bytes.toBytes(ReplicationConstants.ReplicationStatusTool.TIME_FAMILY);
+    this.rowType = Bytes.toBytes(rowType);
+    this.rsID = getUUIDArray(UUID.randomUUID());
+    this.conf = conf;
+
+    String configuredDelay = conf.get(ReplicationConstants.ReplicationStatusTool.REPLICATION_DELAY);
+    String configuredPeriod = conf.get(ReplicationConstants.ReplicationStatusTool.REPLICATION_PERIOD);
+    long delay = (configuredDelay != null)
+      ? new Long(configuredDelay)
+      : ReplicationConstants.ReplicationStatusTool.REPLICATION_DELAY_DEFAULT;
+    long period = (configuredPeriod != null)
+      ? new Long(configuredPeriod)
+      : ReplicationConstants.ReplicationStatusTool.REPLICATION_PERIOD_DEFAULT;
+
+    cachedUpdates = new ConcurrentHashMap<>();
+    setupTimer(delay, period);
+  }
+
+  private byte[] getUUIDArray(UUID uuid) {
+    ByteBuffer byteBuffer = ByteBuffer.wrap(new byte[16]);
+    byteBuffer.putLong(uuid.getMostSignificantBits());
+    byteBuffer.putLong(uuid.getLeastSignificantBits());
+    return byteBuffer.array();
+  }
+
+  /**
+   * Synchronized method used by replication status tool TableUpdate class to add timestamps
+   * from coprocessors.
+   * @param regionName is the key for the Map
+   * @param time is the value for the Map
+   */
+  public synchronized void updateTime(String regionName, long time) {
+    if (!cachedUpdates.containsKey(regionName) || time > cachedUpdates.get(regionName)) {
+      cachedUpdates.put(regionName, time);
+    }
+  }
+
+  protected final byte[] getRowKey(String regionName) {
+    ReplicationStatusKey replicationStatusKey = new ReplicationStatusKey(rowType,
+                                                                         Bytes.toBytes(regionName), rsID);
+    return replicationStatusKey.getKey();
+  }
+
+  public void setupTimer(long delay, long period) {
+    updateTimer = new Timer("UpdateTimer");
+    TimerTask updateTask = new TimerTask() {
+      @Override
+      public void run() {
+        try {
+          if (!tableExists) {
+            createTableIfNotExists(conf);
+            tableExists = true;
+          }
+          if (!cachedUpdates.isEmpty()) {
```

Comment:
It seems `cachedUpdates` is always empty when `setupTimer` is called in the constructor?